File: golangOOP/START_HERE.go
/*
╔═══════════════════════════════════════════════════════════════════════════════╗
║                                                                               ║
║                   GO OOP LEARNING TASKS - GETTING STARTED                     ║
║                                                                               ║
╚═══════════════════════════════════════════════════════════════════════════════╝
*/

package main

import (
	"fmt"
	"strings"
)

func main() {
	printWelcome()
}

func printWelcome() {
	border := strings.Repeat("=", 79)
	
	fmt.Println(border)
	fmt.Println("                  GO OOP LEARNING TASKS - GETTING STARTED")
	fmt.Println(border)
	
	fmt.Println("\nWelcome! You now have everything you need to master Object-Oriented Programming")
	fmt.Println("in Go. Here's how to get started:")
	
	fmt.Println("\n📁 WHAT YOU HAVE")
	fmt.Println(border)
	
	fmt.Println("✅ 10 Task Files (task_01 through task_10)")
	fmt.Println("   - Each contains a detailed problem to solve")
	fmt.Println("   - Progressively increasing difficulty")
	fmt.Println("   - Real-world scenarios to learn from")
	
	fmt.Println("\n✅ 10 Solution Files (in solutions/ folder)")
	fmt.Println("   - Fully explained solutions")
	fmt.Println("   - Detailed comments and documentation")
	fmt.Println("   - Runnable examples with test cases")
	
	fmt.Println("\n✅ README.md")
	fmt.Println("   - Complete guide to using this repository")
	fmt.Println("   - Learning path and recommended approach")
	fmt.Println("   - Go-specific OOP concepts")
	
	fmt.Println("\n✅ QUICK_REFERENCE.md")
	fmt.Println("   - Handy Go syntax reference")
	fmt.Println("   - Common patterns and idioms")
	fmt.Println("   - Quick lookup for concepts")
	
	fmt.Println("\n🚀 QUICK START GUIDE")
	fmt.Println(border)
	
	steps := []string{
		"Read the README:           cat README.md",
		"Check Quick Reference:     cat QUICK_REFERENCE.md",
		"Start with Task 1:         cat task_01_structs_methods.go",
		"Write your solution in the task file",
		"Test your solution:        go run task_01_structs_methods.go",
		"If stuck, check solution:  go run solutions/solution_01_structs_methods.go",
		"Move to Task 2 and repeat!",
	}
	
	for i, step := range steps {
		fmt.Printf("Step %d: %s\n", i+1, step)
	}
	
	fmt.Println("\n📚 LEARNING PATH")
	fmt.Println(border)
	
	printSection("BEGINNER (Start Here!)", []string{
		"Task 1: Structs and Methods              [⭐ Easy]",
		"Task 2: Interfaces                       [⭐⭐ Easy-Medium]",
	})
	
	printSection("INTERMEDIATE", []string{
		"Task 3: Struct Embedding                 [⭐⭐ Medium]",
		"Task 4: Encapsulation                    [⭐⭐⭐ Medium]",
		"Task 5: Interface Composition            [⭐⭐⭐ Medium-Hard]",
	})
	
	printSection("ADVANCED", []string{
		"Task 6: Type Assertions                  [⭐⭐⭐⭐ Hard]",
		"Task 7: Custom Types                     [⭐⭐⭐⭐ Hard]",
		"Task 8: Error Handling                   [⭐⭐⭐⭐ Hard]",
		"Task 9: Generics                         [⭐⭐⭐⭐⭐ Very Hard]",
		"Task 10: Design Patterns                 [⭐⭐⭐⭐⭐ Very Hard]",
	})
	
	fmt.Println("\n💡 TIPS FOR SUCCESS")
	fmt.Println(border)
	
	tips := []string{
		"DON'T RUSH - Take your time, understanding > speed",
		"TRY FIRST - Attempt before checking solutions",
		"TYPE, DON'T COPY - Build muscle memory",
		"USE GO FMT - Format your code: go fmt filename.go",
		"READ ERRORS - Go errors are helpful and clear",
		"THINK IN GO - Embrace composition, not inheritance",
	}
	
	for i, tip := range tips {
		fmt.Printf("%d. %s\n", i+1, tip)
	}
	
	fmt.Println("\n📖 WHAT YOU'LL LEARN")
	fmt.Println(border)
	
	topics := []string{
		"Structs and Methods",
		"Interfaces and Polymorphism",
		"Composition via Embedding",
		"Encapsulation (Exported/Unexported)",
		"Interface Composition",
		"Type Assertions and Type Switches",
		"Custom Types and Methods",
		"Custom Error Types",
		"Generics (Go 1.18+)",
		"Design Patterns in Go",
	}
	
	for _, topic := range topics {
		fmt.Printf("✓ %s\n", topic)
	}
	
	fmt.Println("\n🎯 KEY DIFFERENCES FROM OTHER OOP LANGUAGES")
	fmt.Println(border)
	
	differences := []struct{ from, to string }{
		{"Classes", "Structs"},
		{"Inheritance", "Composition (Embedding)"},
		{"'implements' keyword", "Implicit interface implementation"},
		{"Constructors", "NewTypeName() functions"},
		{"public/private keywords", "Capitalization (Name vs name)"},
		{"Try-Catch exceptions", "Explicit error values"},
	}
	
	for _, diff := range differences {
		fmt.Printf("   %s → %s\n", diff.from, diff.to)
	}
	
	fmt.Println("\n⚡ RUNNING THE CODE")
	fmt.Println(border)
	
	fmt.Println("Run your solution:")
	fmt.Println("   $ go run task_01_structs_methods.go")
	
	fmt.Println("\nRun a solution to see expected output:")
	fmt.Println("   $ go run solutions/solution_01_structs_methods.go")
	
	fmt.Println("\nFormat your code (Go standard):")
	fmt.Println("   $ go fmt task_01_structs_methods.go")
	
	fmt.Println("\nCheck for issues:")
	fmt.Println("   $ go vet task_01_structs_methods.go")
	
	fmt.Println("\n🌟 GO PROVERBS TO REMEMBER")
	fmt.Println(border)
	
	proverbs := []string{
		"Don't communicate by sharing memory, share memory by communicating.",
		"Concurrency is not parallelism.",
		"The bigger the interface, the weaker the abstraction.",
		"Make the zero value useful.",
		"interface{} says nothing.",
		"Gofmt's style is no one's favorite, yet gofmt is everyone's favorite.",
		"A little copying is better than a little dependency.",
		"Clear is better than clever.",
		"Errors are values.",
		"Don't just check errors, handle them gracefully.",
	}
	
	for i, proverb := range proverbs {
		fmt.Printf("%2d. %s\n", i+1, proverb)
	}
	
	fmt.Println("\n" + border)
	fmt.Println("Ready to start? Open task_01_structs_methods.go and begin your journey!")
	fmt.Println()
	fmt.Println("                            HAPPY CODING! 🚀")
	fmt.Println(border)
}

func printSection(title string, items []string) {
	fmt.Printf("\n%s\n", title)
	for _, item := range items {
		fmt.Printf("├── %s\n", item)
	}
}

File: javaThreads/START_HERE.java
/**
 * ═══════════════════════════════════════════════════════════════════════════════
 *                                                                               
 *           JAVA THREADS & CONCURRENCY LEARNING TASKS - GETTING STARTED         
 *                          Target: Java 25 LTS                                  
 *                                                                               
 * ═══════════════════════════════════════════════════════════════════════════════
 * 
 * Welcome! This track will take you from threading basics to modern Java 25
 * concurrency (virtual threads, structured concurrency, scoped values).
 * 
 * 📁 WHAT YOU HAVE
 * ═══════════════════════════════════════════════════════════════════════════════
 * 
 * ✅ 12 Task Files (Task01 through Task12)
 *    - Each contains a detailed problem to solve
 *    - Progressively increasing difficulty
 *    - Real-world concurrency scenarios
 * 
 * ✅ 12 Solution Files (in solutions/ folder)
 *    - Fully explained solutions with detailed annotations
 *    - Executable examples with observable output
 * 
 * ✅ README.md — Complete guide to the learning path
 * ✅ QUICK_REFERENCE.md — Syntax cheat sheet for all concurrency APIs
 * ✅ PROGRESS_TRACKER.md — Track your learning journey
 * 
 * 
 * 🚀 QUICK START GUIDE
 * ═══════════════════════════════════════════════════════════════════════════════
 * 
 * Step 1: Read the README
 *    $ cat README.md
 * 
 * Step 2: Check the Quick Reference
 *    $ cat QUICK_REFERENCE.md
 * 
 * Step 3: Start with Task 1
 *    $ cat Task01ThreadCreation.java
 * 
 * Step 4: Write your solution in the task file
 * 
 * Step 5: Compile and test
 *    $ javac Task01ThreadCreation.java
 *    $ java Task01ThreadCreation
 * 
 * Step 6: If stuck, check the solution
 *    $ javac solutions/Solution01ThreadCreation.java
 *    $ java -cp solutions Solution01ThreadCreation
 * 
 * 
 * 📚 LEARNING PATH
 * ═══════════════════════════════════════════════════════════════════════════════
 * 
 * 🟢 FUNDAMENTALS
 * ├── Task 01: Thread Creation & Lifecycle       [⭐ Beginner]
 * └── Task 02: Synchronization Basics            [⭐⭐ Easy-Medium]
 * 
 * 🟡 INTERMEDIATE
 * ├── Task 03: Deadlock, Livelock & Starvation   [⭐⭐⭐ Medium]
 * ├── Task 04: Volatile & Atomics                [⭐⭐⭐ Medium]
 * ├── Task 05: ReentrantLock & ReadWriteLock      [⭐⭐⭐ Medium]
 * └── Task 06: Thread Communication              [⭐⭐⭐⭐ Medium-Hard]
 * 
 * 🔴 ADVANCED
 * ├── Task 07: Executor Framework                [⭐⭐⭐ Medium]
 * ├── Task 08: CompletableFuture                 [⭐⭐⭐⭐ Hard]
 * ├── Task 09: Synchronization Aids              [⭐⭐⭐⭐ Hard]
 * └── Task 10: Concurrent Collections            [⭐⭐⭐⭐ Hard]
 * 
 * ⚡ MODERN JAVA (Expert)
 * ├── Task 11: ForkJoinPool & Parallel Streams   [⭐⭐⭐⭐ Hard]
 * └── Task 12: Virtual Threads & Java 25 LTS     [⭐⭐⭐⭐⭐ Expert]
 * 
 * 
 * 💡 TIPS FOR CONCURRENCY LEARNING
 * ═══════════════════════════════════════════════════════════════════════════════
 * 
 * 1. THREADING IS NON-DETERMINISTIC
 *    - Run solutions multiple times — output order WILL vary
 *    - This is expected behavior, not a bug
 * 
 * 2. SLOW THINGS DOWN TO SEE BUGS
 *    - Add Thread.sleep() calls to make race conditions visible
 *    - Remove them once you understand the issue
 * 
 * 3. DON'T SKIP DEADLOCK (Task 3)
 *    - It's the #1 concurrency interview question
 *    - Understanding deadlock = understanding thread safety
 * 
 * 4. THINK ABOUT "WHAT IF" SCENARIOS
 *    - What if Thread A runs before Thread B?
 *    - What if both run at the exact same time?
 *    - What if one thread crashes?
 * 
 * 5. USE DEBUGGING TOOLS
 *    - jstack <pid>     — dump thread states
 *    - jconsole          — visual thread monitor
 *    - -XX:+PrintFlagsFinal — JVM thread-related flags
 * 
 * 
 * ═══════════════════════════════════════════════════════════════════════════════
 * 
 * Ready to start? Open Task01ThreadCreation.java and begin your journey!
 * 
 *                         HAPPY THREADING! 🧵✨
 * 
 * ═══════════════════════════════════════════════════════════════════════════════
 */

public class START_HERE {
    public static void main(String[] args) {
        System.out.println("═══════════════════════════════════════════════════════════════════════════════");
        System.out.println("          JAVA THREADS & CONCURRENCY — GETTING STARTED (Java 25 LTS)          ");
        System.out.println("═══════════════════════════════════════════════════════════════════════════════");
        System.out.println();
        System.out.println("Welcome to the Java Threads & Concurrency learning track!");
        System.out.println();
        System.out.println("📚 To get started:");
        System.out.println("   1. Read README.md for the complete learning path");
        System.out.println("   2. Check QUICK_REFERENCE.md for syntax help");
        System.out.println("   3. Open Task01ThreadCreation.java to begin");
        System.out.println();
        System.out.println("🧵 12 Tasks covering:");
        System.out.println("   Tasks 01-02: Thread basics, synchronization");
        System.out.println("   Tasks 03-06: Deadlocks, volatile, locks, wait/notify");
        System.out.println("   Tasks 07-10: Executors, CompletableFuture, concurrent collections");
        System.out.println("   Tasks 11-12: ForkJoin, virtual threads, Java 25 LTS features");
        System.out.println();
        System.out.println("💡 Remember: Threading is non-deterministic — run solutions multiple times!");
        System.out.println();
        System.out.println("═══════════════════════════════════════════════════════════════════════════════");
    }
}
